### ISSUE TITLE
Two ExpressionStatements allowed on one line with no semicolon!

### ISSUE THREAD

**Issue by conartist6** [microsoft/TypeScript#63398]
### 🔎 Search Terms

ASI Automatic Semicolon Insertion

### 🕗 Version & Regression Information

unknown

### 💻 Code

```ts
foo bar;
```

### 🙁 Actual behavior

no parse error

### 🙂 Expected behavior

parse error

**Comment by conartist6** [comment]
Other unexpected results are that this is perfectly valid code according to the TS parser
```ts
class Foo {
  static foo bar baz boom yuck what language is this (){
  }
}
```

**Comment by nmain** [comment]
This doesn't reproduce obviously in 6.0.2.  Please fill out the entire template including a minimal, version numbers, etc.

**Comment by conartist6** [comment]
https://www.typescriptlang.org/play/?#code/MYGwhgzhAEBiD29oG8BQ1oQC5iwS2GgDNFoAjMAJ3LAC9zEBbaATwFdgBraAdwAtc0cADsA5mzCiAptDwwsfOdAAUASjQYAvqk1A

Looks like it repros on 6.0.2 to me

**Comment by nmain** [comment]
That example gives me 10 syntax errors in the playground.  Look at all the squigglies:

<img width="660" height="215" alt="Image" src="https://github.com/user-attachments/assets/09e6bf61-dd61-490f-91d7-8deb529c5f51" />

**Comment by conartist6** [comment]
That example scrambles my brain. If it's just a matter of ignoring some spurious errors, surely that means that in this language omitting those extra symbols is a matter of *semantics* not syntax...?

**Comment by mkantor** [comment]
"Unexpected keyword or identifier" sounds like a syntax error to me. What do you mean by "ignoring some spurious errors"?

**Comment by snarbles2** [comment]
What are you expecting to happen that is not happening?

**Comment by conartist6** [comment]
I was expecting the runtime not to give concrete, definite, consistent (usable) meaning to undefined behavior.

The parse error feels more like a lint rule warning, and the ability to put two statements on one line feels like a fundamental ability of the Typescript language to me...

**Comment by conartist6** [comment]
It's probably only the fact that your parser isn't usually considered the definitive parser for typescript that makes this not really, really dangerous.

It's most likely the Prettier parser that is doing the heavy lifting on defining what syntax is valid Typescript

**Comment by nmain** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/44232.

**Comment by RyanCavanaugh** [comment]
There seems to be some kind of fundamental disconnect in interpretation here, not a bug.

> The parse error feels more like a lint rule warning, and the ability to put two statements on one line feels like a fundamental ability of the Typescript language to me

We can't help that you don't like the vibes of the error. I don't understand how a language would remove the *ability* of you to write invalid syntax, except to define all possible sequences of characters as valid programs, which doesn't seem to be what you're asking for either.

**Comment by conartist6** [comment]
It's more like I'm saying that what TS has done is define all possible sequences of characters as valid programs.

The only thing saving you from the complete realization of that outcome is ESLint and Prettier, which do error on syntax errors.

Take those away and TS has legalized this syntax.

**Comment by snarbles2** [comment]
As was previously pointed out, TS _is_ giving you syntax errors. 

    class Foo {
      static foo bar baz boom yuck what language is this (){
      }
    }

- Member 'foo' implicitly has an 'any' type.
- Member 'bar' implicitly has an 'any' type.
- Member 'baz' implicitly has an 'any' type.
- Member 'boom' implicitly has an 'any' type.
- Member 'yuck' implicitly has an 'any' type.
- Member 'what' implicitly has an 'any' type.
- Member 'language' implicitly has an 'any' type.
- Cannot find name 'is'.
- This expression is not callable.
  - Type 'typeof globalThis' has no call signatures.
- **Unexpected keyword or identifier.**
- **Unexpected keyword or identifier.**
- **Unexpected keyword or identifier.**
- **Unexpected keyword or identifier.**
- **Unexpected keyword or identifier.**
- **Unexpected keyword or identifier.**
- **Unexpected keyword or identifier.**
- **Unexpected token. A constructor, method, accessor, or property was expected.**
- **';' expected.**
- **Declaration or statement expected.**

You write invalid syntax, then you run TSC, then you receive syntax errors. What would you want to happen differently?

**Comment by mkantor** [comment]
@conartist6 you might be looking for [`noEmitOnError`](https://www.typescriptlang.org/tsconfig/#noEmitOnError)? The only way I can make sense of your comments is by assuming you're talking about whether or not JavaScript code is emitted, not whether errors are reported.

**Comment by conartist6** [comment]
Yeah I'm coming at this form the perspective of a computational linguist. What I know about language is that it's an emergent thing that happens when two or more people agree on meanings so much that they can communicate.

If these error messages can be turned off and emitting still happens, then what I'm saying is: that isn't failure to communicate, it's just communication. What the authors have actually done is created a new programming language in which this bowdlerized syntax is fully supported -- or again, would have done if not for Prettier

**Comment by RyanCavanaugh** [comment]
Again, it seems like you're looking for `noEmitOnError`

**Comment by mkantor** [comment]
If a language implementation tells you that something is an error (in this case via documentation, diagnostic messages, and a nonzero exit code from the compiler), then I think it's only fair to consider it to be an error. If TypeScript still had a formal specification then I'm sure this wouldn't be valid syntax according to it, either.

For another example, I think almost everyone would agree that the following Bash program contains invalid syntax, even though  output is still produced when it's executed:

```sh
#!/usr/bin/env bash
echo 'this is output'
|$='but this line is not valid bash syntax
```

---

I'd like to better understand your perspective, but this probably isn't the best forum to gain that understanding. We've talked in public Discord servers before; I'll send you a friend invite over there and if you're interested we can continue chatting.
